app/services/rating_service.rb

```diff
@@ -0,0 +1,21 @@
+class RatingService < BaseService
+  include Dry::Monads[:result]
+
+  def initialize(data)
+    @data = data
+  end
+
+  def call()
+    return Failure(:invalid_data) unless valid?(@data)
+    return Failure(:no_votes) if @data.empty?
+    Success(@data.reduce(0) { |sum, vote| sum + vote.grades } / @data.size.to_f)
+  end
+
+  private
+
+  def valid?(collection)
+    return false unless collection.is_a?(Enumerable)
+    return false unless collection.all?(Vote)
+    true
+  end
```

Comment:
Looks good, but you could use [do-notation](https://dry-rb.org/gems/dry-monads/1.3/do-notation/) here to make it even better.

It'd look something like:

```ruby
def call
  yield validate(@data)
  avg_rating = yield calculate_average_rating(@data)
  Success(avg_rating)
end 

def validate(collection)
  return Failure(:invalid_data) unless collection.is_a?(Enumerable)
  return Failure(:invalid_data) unless collection.all?(Vote)
  return Failure(:no_votes) if collection.empty?
  Success()
end

def calculate_average_rating(data)
  Try { @data.reduce(0) { |sum, vote| sum + vote.grades } / @data.size.to_f }.to_result
end
```